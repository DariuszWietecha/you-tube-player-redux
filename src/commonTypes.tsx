
import { RouterState } from "connected-react-router";
import { Action } from "redux";
import { ThunkAction } from "redux-thunk";

export type DispatchAction<T = void> = ThunkAction<
  Promise<T>,
  IApplicationState,
  void,
  Action
>;

export interface IResource {
  kind: string;
  etag: string;
  snippet: ISnippet;
}
export interface ISearchListSearchResponse extends ISearchResponse {
  items: ISearchResource[];
}

export interface ISearchResource extends IResource {
  id: {
    kind: string;
    videoId: string;
    channelId?: string;
    playlistId?: string;
  };
}
export interface ISearchResponse {
  kind: string;
  etag: string;
  nextPageToken: string;
  prevPageToken?: string;
  regionCode: string;
  pageInfo: {
    totalResults: number;
    resultsPerPage: number;
  };
}

export interface ISnippet {
  publishedAt: Date;
  channelId: string;
  title: string;
  description: string;
  thumbnails: {
    [key: string]: {
      url: string;
      width: number;
      height: number;
    },
  };
  channelTitle: string;
}

// State
export interface IApplicationState {
  query: string;
  router: RouterState;
  searchResults: ISearchResultsState;
  url: string;
}

export interface ISearchResultsState {
  isFetching: boolean;
  items: ISearchResource[];
  nextPageToken?: string;
  video?: ISearchResource;
  index?: number;
}

export function isSearchResource(obj: any): obj is ISearchResource {
  return obj.id && typeof obj.id.videoId !== "undefined";
}
