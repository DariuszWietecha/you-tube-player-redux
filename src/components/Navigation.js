import React from 'react'
import { Button } from 'reactstrap'
import PropTypes from 'prop-types'

const Navigation = ({ searchResults, selectedVideo, selectPrevVideoSetUrl, selectNextVideoSetUrl, unselectVideoSetUrl, query }) => {
  const disabledPrev = selectedVideo.index === 0

  if (selectedVideo.video) {
    return (
      <div className="row justify-content-end">
        <div className="col mb-3 text-right">
          <span className={searchResults.length > 1 ? '' : 'd-none'}>
            <Button
              className="mr-2"
              color="info"
              disabled={disabledPrev}
              onClick={() => selectPrevVideoSetUrl(searchResults, selectedVideo.index)}
              outline>Prev</Button>
            <Button
              className="mr-4"
              color="info"
              onClick={() => selectNextVideoSetUrl(searchResults, selectedVideo.index)}
              outline>Next</Button>
          </span>
          <Button onClick={() => unselectVideoSetUrl(query)} outline>Back</Button>
        </div>
      </div>
    )
  } else {
    return null
  }
}

Navigation.propTypes = {
  searchResults: PropTypes.arrayOf(PropTypes.shape({
    snippet: {
      title: PropTypes.string.isRequired
    }
  }).isRequired).isRequired,
  selectedVideo: {
    snippet: {
      description: PropTypes.string.isRequired,
      title: PropTypes.string.isRequired
    }
  },
  selectPrevVideoSetUrl: PropTypes.func.isRequired,
  selectNextVideoSetUrl: PropTypes.func.isRequired,
  unselectVideoSetUrl: PropTypes.func.isRequired
}

export default Navigation
