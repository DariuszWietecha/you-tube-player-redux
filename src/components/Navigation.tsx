import React from "react";
import { Button } from "reactstrap";
import * as commonTypes from "../commonTypes";

export interface INavigationProps {
  searchResults: commonTypes.ISearchResultsState;
  selectPrevVideoSetUrl: (searchResults: commonTypes.ISearchResource[], index?: number) => commonTypes.DispatchAction;
  selectNextVideoSetUrl: (video: commonTypes.ISearchResource[], index?: number) => commonTypes.DispatchAction;
  unselectVideoSetUrl: (query: string) => commonTypes.DispatchAction;
  query: string;
}

export const Navigation = (props: INavigationProps) => {
  const disabledPrev = props.searchResults.index === 0;

  if (typeof props.searchResults.video !== "undefined") {
    return (
      <div className="row justify-content-end">
        <div className="col mb-3 text-right">
          <span className={props.searchResults.items.length > 1 ? "" : "d-none"}>
            <Button
              id="prev"
              className="mr-2"
              color="info"
              disabled={disabledPrev}
              onClick={() => props.selectPrevVideoSetUrl(props.searchResults.items, props.searchResults.index)}
              outline>Prev</Button>
            <Button
              id="next"
              className="mr-4"
              color="info"
              onClick={() => props.selectNextVideoSetUrl(props.searchResults.items, props.searchResults.index)}
              outline>Next</Button>
          </span>
          <Button
            id="back"
            onClick={() => props.unselectVideoSetUrl(props.query)}
            outline>Back</Button>
        </div>
      </div>
    );
  } else {
    return null;
  }
};
