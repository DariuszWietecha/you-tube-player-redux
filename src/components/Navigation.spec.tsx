import { shallow } from "enzyme";
import React from "react";
import renderer from "react-test-renderer";
import * as commonTypes from "../commonTypes";
import { Navigation } from "./Navigation";

describe("Navigation component", () => {
  const searchResultsItem: commonTypes.ISearchResource = require("../../test/data/searchResultsItem.json");

  const props = {
    query: "query",
    searchResults: {
      index: 0,
      isFetching: false,
      items: [],
    },
    selectNextVideoSetUrl: jest.fn(),
    selectPrevVideoSetUrl: jest.fn(),
    unselectVideoSetUrl: jest.fn(),
  };

  const propsWithVideoItemFirst = {
    ...props,
    searchResults: {
      index: 0,
      isFetching: false,
      items: [searchResultsItem, searchResultsItem],
      video: searchResultsItem,
    },
  };

  const propsWithVideoItemSecond = {
    ...propsWithVideoItemFirst,
    searchResults: {
      ...propsWithVideoItemFirst.searchResults,
      index: 2,
    },
  };

  it("video item not set -> render null", () => {
    const component = renderer.create(<Navigation {...props} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("normal view, video item is the first from the list", () => {
    const component = renderer.create(<Navigation {...propsWithVideoItemFirst} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("normal view, video item is the second from the list", () => {
    const component = renderer.create(<Navigation {...propsWithVideoItemSecond} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("calls the selectNextVideoSetUrl function when the 'Next' is clicked", () => {
    const wrapper = shallow(<Navigation {...propsWithVideoItemFirst} />);
    wrapper.find("#next").simulate("click");
    expect(propsWithVideoItemFirst.selectNextVideoSetUrl).toHaveBeenCalled();
  });
});
