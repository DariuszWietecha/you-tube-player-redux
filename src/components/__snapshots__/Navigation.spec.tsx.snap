// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component normal view, video item is the first from the list 1`] = `
<div
  className="row justify-content-end"
>
  <div
    className="col mb-3 text-right"
  >
    <span
      className=""
    >
      <button
        aria-label={null}
        className="mr-2 btn btn-outline-info disabled"
        disabled={true}
        id="prev"
        onClick={[Function]}
        type="button"
      >
        Prev
      </button>
      <button
        aria-label={null}
        className="mr-4 btn btn-outline-info"
        id="next"
        onClick={[Function]}
        type="button"
      >
        Next
      </button>
    </span>
    <button
      aria-label={null}
      className="btn btn-outline-secondary"
      id="back"
      onClick={[Function]}
      type="button"
    >
      Back
    </button>
  </div>
</div>
`;

exports[`App Component normal view, video item is the second from the list 1`] = `
<div
  className="row justify-content-end"
>
  <div
    className="col mb-3 text-right"
  >
    <span
      className=""
    >
      <button
        aria-label={null}
        className="mr-2 btn btn-outline-info"
        disabled={false}
        id="prev"
        onClick={[Function]}
        type="button"
      >
        Prev
      </button>
      <button
        aria-label={null}
        className="mr-4 btn btn-outline-info"
        id="next"
        onClick={[Function]}
        type="button"
      >
        Next
      </button>
    </span>
    <button
      aria-label={null}
      className="btn btn-outline-secondary"
      id="back"
      onClick={[Function]}
      type="button"
    >
      Back
    </button>
  </div>
</div>
`;

exports[`App Component video item not set -> render null 1`] = `null`;
